File: examples/langgraph-code-agent/agent.py
"""
Multi-agent code generator using LangGraph.

A pipeline of 4 agents:
  Plan -> Code -> Execute -> Summarize

With automatic retry on execution failure (up to 2 retries).
"""

import logging
import os
import subprocess
import sys
import tempfile
from typing import Any, TypedDict

from langchain.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI
from langgraph.graph import END, StateGraph
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# State
# ---------------------------------------------------------------------------


class AgentState(TypedDict):
    query: str
    is_coding: bool
    plan: str
    code: str
    execution_result: dict[str, Any]
    response: str | None
    retry_count: int
    max_retries: int
    last_summary: str


# ---------------------------------------------------------------------------
# Plan agent
# ---------------------------------------------------------------------------


class PlanResponse(BaseModel):
    is_coding: bool = Field(description="Whether the query is coding-related")
    plan: str | None = Field(default=None, description="Plan for coding tasks")
    response: str | None = Field(default=None, description="Direct response for non-coding queries")


def plan_node(state: AgentState) -> dict:
    llm = ChatOpenAI(model="gpt-4o", temperature=0)
    prompt = ChatPromptTemplate.from_messages(
        [
            (
                "system",
                "You are a planning agent. Determine if the query is coding-related. "
                "If yes, set is_coding=true and provide a concise plan. "
                "If not, set is_coding=false and answer directly. "
                "If a previous summary is provided, learn from failures and improve.",
            ),
            ("human", "{query}"),
        ]
    )

    query = state["query"]
    if state["last_summary"]:
        query = f"{query}\n\nPrevious attempt summary:\n{state['last_summary']}"

    chain = prompt | llm.with_structured_output(PlanResponse)
    result = chain.invoke({"query": query})

    return {
        "is_coding": result.is_coding,
        "plan": result.plan or "",
        "response": result.response,
    }


# ---------------------------------------------------------------------------
# Code agent
# ---------------------------------------------------------------------------


def code_node(state: AgentState) -> dict:
    llm = ChatOpenAI(model="gpt-4o", temperature=0)
    prompt = ChatPromptTemplate.from_messages(
        [
            (
                "system",
                "You are a Python coding agent. Write clean, executable Python code. "
                "Include necessary imports. Return ONLY the code, no explanations. "
                "If historical context is provided, avoid repeating previous mistakes.",
            ),
            ("human", "{query}\n\nPlan: {plan}\n\nContext: {context}\n\nWrite the Python code."),
        ]
    )

    chain = prompt | llm
    response = chain.invoke(
        {
            "query": state["query"],
            "plan": state["plan"],
            "context": state["last_summary"],
        }
    )

    code = response.content.strip()
    if code.startswith("```python"):
        code = code[9:]
    elif code.startswith("```"):
        code = code[3:]
    if code.endswith("```"):
        code = code[:-3]

    return {"code": code.strip()}


# ---------------------------------------------------------------------------
# Execution agent
# ---------------------------------------------------------------------------


def execute_node(state: AgentState) -> dict:
    try:
        with tempfile.NamedTemporaryFile(mode="w", suffix=".py", delete=False) as f:
            f.write(state["code"])
            tmp = f.name

        result = subprocess.run(
            [sys.executable, tmp],
            capture_output=True,
            text=True,
            timeout=30,
        )
        return {
            "execution_result": {
                "success": result.returncode == 0,
                "stdout": result.stdout,
                "stderr": result.stderr,
            }
        }
    except subprocess.TimeoutExpired:
        return {
            "execution_result": {"success": False, "stdout": "", "stderr": "Timed out after 30s"}
        }
    except Exception as e:
        return {"execution_result": {"success": False, "stdout": "", "stderr": str(e)}}
    finally:
        try:
            os.unlink(tmp)
        except Exception:
            pass


# ---------------------------------------------------------------------------
# Summarize agent
# ---------------------------------------------------------------------------


def summarize_node(state: AgentState) -> dict:
    if not state["is_coding"]:
        return {"response": state["response"]}

    llm = ChatOpenAI(model="gpt-4o", temperature=0)
    prompt = ChatPromptTemplate.from_messages(
        [
            (
                "system",
                "Summarize the coding task result. Be concise and helpful. "
                "Show the output or explain errors.",
            ),
            (
                "human",
                "Query: {query}\nPlan: {plan}\nCode:\n```python\n{code}\n```\n"
                "Success: {success}\nOutput: {stdout}\nError: {stderr}\n"
                "Retry #{retry_count}",
            ),
        ]
    )

    er = state["execution_result"]
    chain = prompt | llm
    response = chain.invoke(
        {
            "query": state["query"],
            "plan": state["plan"],
            "code": state["code"],
            "success": er.get("success", False),
            "stdout": er.get("stdout", ""),
            "stderr": er.get("stderr", ""),
            "retry_count": state["retry_count"],
        }
    )

    return {
        "response": response.content,
        "last_summary": response.content,
        "retry_count": state["retry_count"] + (0 if er.get("success") else 1),
    }


# ---------------------------------------------------------------------------
# Graph
# ---------------------------------------------------------------------------


def _should_code(state: AgentState) -> str:
    return "code" if state["is_coding"] else "end"


def _should_retry(state: AgentState) -> str:
    er = state["execution_result"]
    if state["is_coding"] and not er.get("success") and state["retry_count"] < state["max_retries"]:
        return "retry"
    return "end"


def build_graph():
    wf = StateGraph(AgentState)
    wf.add_node("planning", plan_node)
    wf.add_node("coding", code_node)
    wf.add_node("execute", execute_node)
    wf.add_node("summarize", summarize_node)

    wf.set_entry_point("planning")
    wf.add_conditional_edges("planning", _should_code, {"code": "coding", "end": "summarize"})
    wf.add_edge("coding", "execute")
    wf.add_edge("execute", "summarize")
    wf.add_conditional_edges("summarize", _should_retry, {"retry": "planning", "end": END})

    return wf.compile()


def process_query(query: str) -> str:
    graph = build_graph()
    result = graph.invoke(
        {
            "query": query,
            "is_coding": False,
            "plan": "",
            "code": "",
            "execution_result": {},
            "response": None,
            "retry_count": 0,
            "max_retries": 2,
            "last_summary": "",
        }
    )
    return result["response"]

File: examples/typescript/langchain/agent.ts
/**
 * LangGraph Code Agent — TraceRoot Observability
 *
 * A LangGraph-powered agent that plans, writes, and executes Python code
 * to answer user queries. Instrumented with TraceRoot via TraceRoot.initialize().
 *
 * langchain: patches the LangChain callback manager to trace all chain/node/LLM spans.
 *
 * Graph: START → plan → code → execute → [retry? → code] → summarize → END
 *
 * Env vars required: OPENAI_API_KEY, TRACEROOT_API_KEY
 *
 * Run:
 *   pnpm demo
 */

import 'dotenv/config';
import { exec } from 'child_process';
import { promisify } from 'util';
import { writeFileSync, unlinkSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';

import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage, SystemMessage } from '@langchain/core/messages';
import { Annotation, END, START, StateGraph } from '@langchain/langgraph';
import * as lcCallbackManager from '@langchain/core/callbacks/manager';
import { TraceRoot, observe } from '@traceroot/sdk';

// ── TraceRoot setup ───────────────────────────────────────────────────────────
// langchain: patches the LangChain callback manager to trace chain/node/LLM spans.
// Do NOT pass openAI here — LangChain instrumentation already captures LLM spans;
// adding the raw OpenAI instrumentation creates duplicate spans.
TraceRoot.initialize({
  instrumentModules: { langchain: lcCallbackManager },
});
console.log('[Observability: TraceRoot]');

// ── LLM ───────────────────────────────────────────────────────────────────────
const llm = new ChatOpenAI({ model: 'gpt-4o-mini', temperature: 0 });

// ── Python code execution ─────────────────────────────────────────────────────
const execAsync = promisify(exec);

async function runPythonCode(code: string): Promise<string> {
  const tmpFile = join(tmpdir(), `agent_code_${Date.now()}.py`);
  try {
    writeFileSync(tmpFile, code, 'utf8');
    const { stdout, stderr } = await execAsync(`python3 "${tmpFile}"`, { timeout: 30_000 });
    if (stderr.trim()) return `ERROR: ${stderr.trim()}`;
    return stdout.trim() || '(no output)';
  } catch (e: unknown) {
    const err = e as { stderr?: string; message?: string };
    return `ERROR: ${err.stderr?.trim() ?? err.message ?? String(e)}`;
  } finally {
    try {
      unlinkSync(tmpFile);
    } catch {
      /* ignore */
    }
  }
}

// ── LangGraph state ───────────────────────────────────────────────────────────
const AgentState = Annotation.Root({
  query: Annotation<string>({ reducer: (_, v) => v, default: () => '' }),
  plan: Annotation<string>({ reducer: (_, v) => v, default: () => '' }),
  code: Annotation<string>({ reducer: (_, v) => v, default: () => '' }),
  executionResult: Annotation<string>({ reducer: (_, v) => v, default: () => '' }),
  errorSummary: Annotation<string>({ reducer: (_, v) => v, default: () => '' }),
  summary: Annotation<string>({ reducer: (_, v) => v, default: () => '' }),
  retries: Annotation<number>({ reducer: (_, v) => v, default: () => 0 }),
});
type State = typeof AgentState.State;

// ── Graph nodes ───────────────────────────────────────────────────────────────
async function planNode(state: State): Promise<Partial<State>> {
  const retryCtx = state.errorSummary
    ? `\nPrevious attempt failed: ${state.errorSummary}\nPlease adjust the approach.`
    : '';
  const response = await llm.invoke([
    new SystemMessage(
      'You are a code planning assistant. Given a user query, produce a concise step-by-step plan for Python code that solves it.',
    ),
    new HumanMessage(`Query: ${state.query}${retryCtx}`),
  ]);
  const plan = response.content as string;
  console.log('\n[Plan]\n' + plan);
  return { plan };
}

async function codeNode(state: State): Promise<Partial<State>> {
  const retryCtx = state.errorSummary
    ? `\nPrevious code failed with: ${state.errorSummary}\nPlease fix the issues.`
    : '';
  const response = await llm.invoke([
    new SystemMessage(
      'You are a Python code generator. Generate clean, executable Python code based on the plan. ' +
        'Output ONLY raw Python code — no markdown fences, no explanations.',
    ),
    new HumanMessage(`Plan:\n${state.plan}\n\nQuery: ${state.query}${retryCtx}`),
  ]);
  let code = response.content as string;
  code = code
    .replace(/^```(?:python)?\n?/m, '')
    .replace(/```\s*$/m, '')
    .trim();
  console.log('\n[Code]\n' + code);
  return { code };
}

async function executeNode(state: State): Promise<Partial<State>> {
  console.log('\n[Executing Python code...]');
  const executionResult = await runPythonCode(state.code);
  const preview = executionResult.slice(0, 300) + (executionResult.length > 300 ? '...' : '');
  console.log('[Result] ' + preview);
  if (executionResult.startsWith('ERROR:')) {
    return { executionResult, errorSummary: executionResult, retries: state.retries + 1 };
  }
  return { executionResult, errorSummary: '' };
}

async function summarizeNode(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage(
      "You are a helpful assistant. Summarize the answer to the user's query based on the Python code and its execution results.",
    ),
    new HumanMessage(
      `Query: ${state.query}\n\nCode:\n${state.code}\n\nExecution Result:\n${state.executionResult}`,
    ),
  ]);
  return { summary: response.content as string };
}

function shouldRetry(state: State): 'code' | 'summarize' {
  if (state.executionResult.startsWith('ERROR:') && state.retries <= 2) {
    console.log(`\n[Retrying... attempt ${state.retries}/2]`);
    return 'code';
  }
  return 'summarize';
}

// ── Build graph ───────────────────────────────────────────────────────────────
const app = new StateGraph(AgentState)
  .addNode('planner', planNode)
  .addNode('coder', codeNode)
  .addNode('executor', executeNode)
  .addNode('summarizer', summarizeNode)
  .addEdge(START, 'planner')
  .addEdge('planner', 'coder')
  .addEdge('coder', 'executor')
  .addConditionalEdges('executor', shouldRetry, { code: 'coder', summarize: 'summarizer' })
  .addEdge('summarizer', END)
  .compile();

// ── Run agent ─────────────────────────────────────────────────────────────────
async function runAgent(query: string): Promise<string> {
  return observe({ name: 'langchain_agent', type: 'agent' }, async () => {
    const result = await app.invoke({ query, retries: 0 });
    return result.summary;
  });
}

// ── Demo ──────────────────────────────────────────────────────────────────────
const DEMO_QUERIES = [
  'Calculate the first 20 Fibonacci numbers and print them.',
  'Find all prime numbers up to 100 using the Sieve of Eratosthenes and print the list.',
  'Compute 15! (factorial of 15) and express it in scientific notation.',
];

async function main() {
  try {
    await observe({ name: 'demo_session' }, async () => {
      console.log('='.repeat(60));
      console.log('LangGraph Code Agent — Demo (TraceRoot)');
      console.log('='.repeat(60));

      for (let i = 0; i < DEMO_QUERIES.length; i++) {
        const query = DEMO_QUERIES[i];
        console.log(`\n${'='.repeat(60)}`);
        console.log(`Query ${i + 1}: ${query}`);
        console.log('='.repeat(60));
        const summary = await runAgent(query);
        console.log('\nAgent: ' + summary);
        console.log();
      }
    });
  } finally {
    await TraceRoot.shutdown();
    console.log('[Traces exported]');
  }
}

main().catch(console.error);
